### PR TITLE
core, netty, okhttp: implement new logic for nameResolverFactory API in channelBuilder

### PR DESCRIPTION
**Pull request by sanjaypujare**
fix ManagedChannelImpl to use NameResolverRegistry instead of NameResolverFactory fix the ManagedChannelImplBuilder and remove nameResolverFactory

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java [hunk 1]
```diff
@@ -259,8 +264,19 @@ private static String getAuthorityFromAddress(SocketAddress address) {
...
+  @SuppressWarnings("unchecked")
   public NettyChannelBuilder channelType(Class<? extends Channel> channelType) {
     checkNotNull(channelType, "channelType");
+    try {
+      Method method = channelType.getMethod("remoteAddress");
+      checkNotNull(method, "method for remoteAddress");
+      Class<?> returnType = method.getReturnType();
+      checkState(SocketAddress.class.isAssignableFrom(returnType),
+          "Unexpected return type:" + returnType.getSimpleName());
+      this.transportSocketType = (Class<? extends SocketAddress>) returnType;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java:276]: This only mutates one of the code paths. If the user calls `channelFactory()` directly, then the address types get in an inconsistent state. For the factory, it seems we need users to pass in the supported types for us to do any checking.

I feel like we can't fix up all cases now (or maybe ever), so there needs to be an opt-out of the address compat check. That could be "return null from getSupportedSocketAddressTypes()". We'd use it in channelFactory() and if vanilla SocketAddress is detected here. And we'd add a new overload of channelType()/channelFactory() where you pass in the supported types to skip the reflection. We'd then use that overload in UdsNettyChannelBuilder.

**Comment by ejona86** [comment]
FYI @markb74, we're adding in checks to notice earlier if the NameResolver returns addresses that aren't supported by the transport. A mis-match was causing pretty confusing errors. I think we should be able to migrate any binder stuff easily, but if not, we should be able to punch holes to disable the check.

File: api/src/main/java/io/grpc/NameResolverRegistry.java [hunk 2]
```diff
@@ -159,6 +160,22 @@ static List<Class<?>> getHardCodedClasses() {
     return Collections.unmodifiableList(list);
   }
 
+  public static NameResolverProvider getNameResolverProvider(
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply api/src/main/java/io/grpc/NameResolverRegistry.java]: good point. Fixed

File: alts/src/test/java/io/grpc/alts/internal/GoogleDefaultProtocolNegotiatorTest.java [hunk 3]
```diff
@@ -61,7 +61,7 @@ public abstract static class HandlerSelectionTest {
 
       @Override
       public Channel getObject() {
-        return InProcessChannelBuilder.forName("test").build();
+        return InProcessChannelBuilder.forName("inprocess:///test").build();
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply alts/src/test/java/io/grpc/alts/internal/GoogleDefaultProtocolNegotiatorTest.java]: Yes, it's not necessary - perhaps overzealous change on my behalf. Fixed

**Comment by sanjaypujare** [comment]
Addressed all comments. PTAL

File: netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java (hunk 1 above)
Comment (sanjaypujare) [reply netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java]: Addressed.

**Comment by sanjaypujare** [comment]
Friendly ping @ejona86

**Comment by ejona86** [comment]
@sanjaypujare, I just pushed a commit. Take a look at it and tell me how you feel about it.

File: api/src/main/java/io/grpc/NameResolverRegistry.java [hunk 4]
```diff
@@ -59,22 +58,16 @@ public final class NameResolverRegistry {
...
-    if (nameResolverProvider == null) {
-      nameResolverProvider = providers().get(
-          asFactory().getDefaultScheme());
-    }
-    return nameResolverProvider;
+  public synchronized String getDefaultScheme() {
+    return defaultScheme;
   }
 
+  public NameResolverProvider get(String scheme) {
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/NameResolverRegistry.java:65]: Should we call it `getForScheme`?

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 5]
```diff
@@ -708,62 +708,58 @@ public CallTracer create() {
...
+      // For "localhost:8080" this would likely cause provider to be null, because "localhost" is
+      // parsed as the scheme. Will hit the next case and try "dns:///localhost:8080".
+      provider = nameResolverRegistry.get(targetUri.getScheme());
     }
 
-    // If we reached here, the targetUri couldn't be used.
-    if (!URI_PATTERN.matcher(target).matches()) {
+    if (provider == null && !URI_PATTERN.matcher(target).matches()) {
       // It doesn't look like a URI target. Maybe it's an authority string. Try with the default
       // scheme from the factory.
```
Comment (sanjaypujare) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:730]: s/from the factory/from the registry/

**Review by sanjaypujare — COMMENTED** [review]
For the last commit from @ejona86 : looks good. Just have 2 comments. Will discuss offline and get them resolved

File: api/src/main/java/io/grpc/NameResolverRegistry.java (hunk 4 above)
Comment (sanjaypujare) [reply api/src/main/java/io/grpc/NameResolverRegistry.java]: renamed

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 5 above)
Comment (sanjaypujare) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: done

**Review by ejona86 — APPROVED** [review]

**Pull request merged**